nixos/modules/services/monitoring/osquery.nix

```diff
@@ -0,0 +1,77 @@
+{ config, lib, pkgs, ... }:
+
+with builtins;
+with lib;
+
+let
+  cfg = config.services.osquery;
+  runtimeValue = flag: default:
+    cfg.config.options.${flag}
+      or cfg.flags.${flag}
+      or default;
+in
+{
+  options.services.osquery = {
+    enable = mkEnableOption (mdDoc "osqueryd daemon");
+
+    config = mkOption {
+      default = { };
+      description = mdDoc ''
+        Configuration to be written to the osqueryd JSON configuration file.
+        To understand the configuration format, refer to https://osquery.readthedocs.io/en/stable/deployment/configuration/#configuration-components.
+      '';
+      example = {
+        options.utc = false;
+      };
+      type = types.attrs;
+    };
+
+    flagfile = mkOption {
+      default = "/etc/osquery/osquery.flags";
+      description = mdDoc ''
+        Path to the flagfile used to provide CLI flags to osqueryd.
+        For more information, refer to https://osquery.readthedocs.io/en/stable/installation/cli-flags/#flagfile.
+      '';
+      type = types.path;
+    };
+    flags = mkOption {
+      default = { };
+      description = mdDoc ''
+        Attribute set of flag names and values to be written to the osqueryd flagfile.
+        For more information, refer to https://osquery.readthedocs.io/en/stable/installation/cli-flags.
+      '';
+      example = {
+        config_refresh = "10";
+      };
+      type = types.attrsOf types.str;
+    };
+  };
+
+  config = mkIf cfg.enable {
+    environment.etc = {
+      "osquery/osquery.conf".text = toJSON cfg.config;
+      "osquery/osquery.flags".text = concatStringsSep "\n" (mapAttrsToList (name: value: "--${name}=${value}") cfg.flags);
+    };
+    environment.systemPackages = [ pkgs.osquery ];
+    systemd.services.osqueryd = {
+      after = [ "network.target" "syslog.service" ];
+      description = "The osquery daemon";
+      preStart = ''
+        mkdir -p $(dirname ${escapeShellArg cfg.flagfile})
+
+        mkdir -p ${escapeShellArg (runtimeValue "logger_path" "/var/log/osquery") }
+
+        mkdir -p $(dirname ${escapeShellArg (runtimeValue "database_path" "/var/osquery/osquery.db")})
```

Comment:
Might like `database_path` and `pidfile` and `logger_path` to be explicit service config options.

Reply:
Yeah, I know these were present in the original module and I have purposefully removed them because I was unsure about committing to maintaining options for all flags as describe in my earlier wall of text.

The old module had an `extraConfig` option for passing configuration that was not explicitly defined as modules options which I suppose we could adapt instead. It would need to facilitate passing both extra configuration file config and CLI options to make the software completely configurable. 

Let me know what you'd prefer given my explanation in my earlier comment. I'm absolutely up for changing this if you would prefer an subset of first class module options over the current interface.

Reply:
The reason I think `extraConfig` isn't appropriate here is basically just that we need to care about them during `preStart`. We could try to pull them out of `extraConfig`, but at that point we should just have first-class config options for these.

Reply:
I resolve this discussion since @znewman01 approved the MR.